desktop/notification/notificationtest/fdo.go

```diff
@@ -0,0 +1,189 @@
+	AppName string
+	Icon    string
+	Summary string
+	Body    string
+	Actions []string
+	Hints   map[string]dbus.Variant
+	Expires int32
+}
+
+type FdoServer struct {
+	conn *dbus.Conn
+	err  *dbus.Error
+
+	mu            sync.Mutex
+	lastID        uint32
+	notifications map[uint32]*FdoNotification
+}
+
+func NewFdoServer() (*FdoServer, error) {
+	conn, err := dbusutil.SessionBusPrivate()
+	if err != nil {
+		return nil, err
+	}
+
+	server := new(FdoServer)
+	server.conn = conn
+	server.notifications = make(map[uint32]*FdoNotification)
+	conn.Export(fdoApi{server}, fdoObjectPath, fdoInterface)
+
+	reply, err := conn.RequestName(fdoBusName, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		conn.Close()
+		return nil, fmt.Errorf("cannot obtain bus name %q", fdoBusName)
+	}
+	return server, nil
+}
+
+func (server *FdoServer) Stop() error {
+	if _, err := server.conn.ReleaseName(fdoBusName); err != nil {
+		return err
+	}
+	return server.conn.Close()
+}
+
+func (server *FdoServer) SetError(err *dbus.Error) {
+	server.err = err
+}
+
+func (server *FdoServer) Get(id uint32) *FdoNotification {
+	server.mu.Lock()
+	defer server.mu.Unlock()
+
+	return server.notifications[id]
+}
+
+func (server *FdoServer) GetAll() []*FdoNotification {
+	server.mu.Lock()
+	defer server.mu.Unlock()
+
+	notifications := make([]*FdoNotification, 0, len(server.notifications))
+	for _, n := range server.notifications {
+		notifications = append(notifications, n)
+	}
+	sort.Slice(notifications, func(i, j int) bool {
+		return notifications[i].ID < notifications[j].ID
+	})
+	return notifications
+}
+
+func (server *FdoServer) Close(id, reason uint32) error {
+	server.mu.Lock()
+	defer server.mu.Unlock()
+
+	if _, ok := server.notifications[id]; !ok {
+		return fmt.Errorf("No such notification: %d", id)
+	}
+	delete(server.notifications, id)
+	return server.conn.Emit(fdoObjectPath, fdoInterface+".NotificationClosed", id, reason)
+}
+
+func (server *FdoServer) InvokeAction(id uint32, actionKey string) error {
+	return server.conn.Emit(fdoObjectPath, fdoInterface+".ActionInvoked", id, actionKey)
+}
+
+type fdoApi struct {
+	server *FdoServer
+}
+
+func (a fdoApi) GetCapabilities() ([]string, *dbus.Error) {
+	if a.server.err != nil {
+		return nil, a.server.err
+	}
+
+	return []string{"cap-foo", "cap-bar"}, nil
+}
+
+func (a fdoApi) Notify(appName string, replacesID uint32, icon, summary, body string, actions []string, hints map[string]dbus.Variant, expires int32) (uint32, *dbus.Error) {
+	if a.server.err != nil {
+		return 0, a.server.err
+	}
+
+	a.server.mu.Lock()
+	defer a.server.mu.Unlock()
+
+	a.server.lastID += 1
+	notification := &FdoNotification{
+		ID:      a.server.lastID,
+		AppName: appName,
+		Icon:    icon,
+		Summary: summary,
+		Body:    body,
+		Actions: actions,
+		Hints:   hints,
+		Expires: expires,
+	}
+	if replacesID != 0 {
+		delete(a.server.notifications, replacesID)
+	}
```

Comment:
Nitpick, it's not really relevant. To avoid unbounded growth of the map, it would be good to recycle the map here when `len(a.server.notifications) == 0`. We follow this pattern in a few places.

Reply:
This server will generally exist for the duration of a single test case.  Is it really worth adding to the complexity when it is quite possible that the test will complete before the GC runs anyway?

Reply:
No, as I said this is fine. I just wanted to raise this as something that was not immediately intuitive to me. For test code this is fine.